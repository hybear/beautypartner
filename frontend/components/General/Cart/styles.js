import styled from 'styled-components';
import { Button as GButton } from '../Form';

export const Button = styled(GButton)`
  && {
    padding: 15px;
    font-family: QuadraSans-Bold;
    letter-spacing: 0.1em;
  }
`;

export const CartStyles = styled.div`
  background: white;
  box-shadow: 0 0 10px 3px rgba(0, 0, 0, 0.2);
  bottom: 0;
  display: grid;
  grid-template-rows: auto 1fr auto;
  grid-template-columns: minmax(15%, 400px);
  height: 100%;
  padding: 10px;
  position: fixed;
  right: 0;
  transform: translateX(105%);
  transition: all 0.3s cubic-bezier(0.27, 0.91, 0.38, 1.02);
  top: 0;
  z-index: 20;
  ${({ open }) => open && `transform: translateX(0);`};
`;

export const CartItemList = styled.ul`
  max-height: 100%;
  overflow: auto;
`;

export const Total = styled.p`
  font-size: 1.25em;
  font-family: QuadraSans-Bold;
  margin-bottom: 10px;
  padding: 10px;
`;

export const CartItemStyles = styled.li`
  position: relative;
  display: grid;
  grid-template:
    'image title title' auto
    'image value value' auto
    /* "image total total" auto */
    / 30% 50% 20%;
  align-items: center;
  border-bottom: 1px solid ${({ theme }) => theme.color.lightGrey};
  padding: 20px 0;
  .image {
    max-width: 100%;
    grid-area: image;
  }
  .title {
    grid-area: title;
  }
  .listPrice {
    color: ${({ theme }) => theme.color.lightGrey};
    margin-left: 5px;
  }
  .bestPrice {
    margin-left: 5px;
  }
  .value {
    grid-area: value;
    display: flex;
  }
  .total {
    grid-area: total;
  }
`;

export const CloseButton = styled.span`
  color: ${({ theme }) => theme.color.black};
  font-size: 3em;
  line-height: initial;
  border: 0;
  position: absolute;
  z-index: 2;
  cursor: pointer;
`;
