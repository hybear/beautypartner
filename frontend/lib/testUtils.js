import casual from 'casual';

// seed it so we get consistent results
casual.seed(777);

const fakeItem = () => ({
  __typename: 'Item',
  id: 'abc123',
  listPrice: 180000,
  bestPrice: 155200,
  category: 100,
  reference: '15523',
  title: 'LA MER The Moisturizing Soft Cream',
  description:
    'An ultra-soft face moisturizer that nurtures for beautiful, glowing radiance, leaving skin looking naturally brightened, hydrated, and restored.',
  image: 'laMer.jpg',
  largeImage: 'largelaMer.jpg',
});

const fakeUser = () => ({
  __typename: 'User',
  id: '4234',
  name: casual.name,
  email: casual.email,
  permissions: ['ADMIN'],
  birthday: '1996-03-08T14:18:00.000Z',
  creditCards: [],
  phone: casual.phone,
  balance: 1800,
  orders: [],
  cart: [],
});

const fakeOrderItem = () => ({
  __typename: 'OrderItem',
  id: casual.uuid,
  image: `${casual.word}.jpg`,
  title: casual.words(),
  listPrice: 4234,
  bestPrice: 4234,
  quantity: 1,
  description: casual.words(),
});

const fakeOrder = () => ({
  __typename: 'Order',
  id: 'ord123',
  charge: 'ch_123',
  total: 400,
  cashback: 5,
  paymentMethod: 'stripe',
  items: [fakeOrderItem(), fakeOrderItem()],
  status: ['APROVED'],
  createdAt: '2020-02-17T01:52:31.269Z',
  user: fakeUser(),
});

const fakeCartItem = overrides => ({
  __typename: 'CartItem',
  id: 'omg123',
  quantity: 3,
  item: fakeItem(),
  user: fakeUser(),
  ...overrides,
});

// Fake LocalStorage
class LocalStorageMock {
  constructor() {
    this.store = {};
  }

  clear() {
    this.store = {};
  }

  getItem(key) {
    return this.store[key] || null;
  }

  setItem(key, value) {
    this.store[key] = value.toString();
  }

  removeItem(key) {
    delete this.store[key];
  }
}

export { LocalStorageMock, fakeItem, fakeUser, fakeCartItem, fakeOrder, fakeOrderItem };
