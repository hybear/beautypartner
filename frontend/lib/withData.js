import withApollo from 'next-with-apollo';
import ApolloClient from 'apollo-boost';
import { endpoint, prodEndpoint } from '../config';

function createClient({ headers }) {
  return new ApolloClient({
    uri: process.env.NODE_ENV === 'development' ? endpoint : prodEndpoint,
    request: operation => {
      operation.setContext({
        fetchOptions: {
          credentials: 'include',
        },
        headers,
      });
    },
    clientState: {
      resolvers: {
        Mutation: {
          // Logout Method to AddToCart
        },
      },
      defaults: {},
    },
  });
}

export default withApollo(createClient);
