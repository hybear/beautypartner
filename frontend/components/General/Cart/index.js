import React, { useState } from 'react';
import { Query, Mutation } from 'react-apollo';
import { adopt } from 'react-adopt';
import gql from 'graphql-tag';
import User from '../../User';
import formatMoney from '../../../lib/formatMoney';
import calcTotalPrice from '../../../lib/calcTotalPrice';
import RemoveFromCart from '../CartRemoveItem';
import Link from 'next/link';
import Stripe from '../../Checkout/Integrations/Stripe';

// UI
import { Button, CartStyles, CartItemList, CartItemStyles, CloseButton, Total } from './styles';
import { Title } from '../';

const LOCAL_STATE_QUERY = gql`
  query {
    cartOpen @client
  }
`;

const TOGGLE_CART_MUTATION = gql`
  mutation {
    toggleCart @client
  }
`;

const Composed = adopt({
  user: ({ render }) => <User>{render}</User>,
  toggleCart: ({ render }) => <Mutation mutation={TOGGLE_CART_MUTATION}>{render}</Mutation>,
  localState: ({ render }) => <Query query={LOCAL_STATE_QUERY}>{render}</Query>,
});

const CartItem = cartItem => {
  if (!cartItem.item)
    return (
      <CartItemStyles>
        <p>This Item has been removed</p>
      </CartItemStyles>
    );
  return (
    <CartItemStyles>
      <RemoveFromCart id={cartItem.id} />
      <img className="image" src={cartItem.item.image}></img>
      <p className="title">{cartItem.item.title}</p>
      <em className="value">
        {cartItem.quantity} &times;
        {cartItem.item.listPrice != cartItem.item.bestPrice && (
          <del className="listPrice">{formatMoney(cartItem.item.listPrice)}</del>
        )}
        <p className="bestPrice">{formatMoney(cartItem.item.bestPrice)}</p>
      </em>
      {/* <p className="total">{formatMoney(cartItem.item.bestPrice * cartItem.quantity)}</p> */}
    </CartItemStyles>
  );
};

const CartContainer = () => {
  return (
    <Composed>
      {({ user, toggleCart, localState }) => {
        const me = user.data.me;
        // if(!me) return null;
        return (
          <>
            <CartStyles open={localState.data.cartOpen}>
              <header>
                <CloseButton onClick={toggleCart}>&times;</CloseButton>
                <Title>{me && me.name.split(' ', 1) + "'s"} Cart</Title>
                {me && (
                  <p>
                    {me.cart.length > 0 ? `You have ${me.cart.length} products in your cart` : `Your cart is empty`}.
                  </p>
                )}
              </header>
              <CartItemList>{me && me.cart.map((cartItem, i) => <CartItem {...cartItem} key={i} />)}</CartItemList>
              <footer>
                <Total>Total: {me && formatMoney(calcTotalPrice(me.cart))}</Total>
                {/* <Link href="/checkout">
                                    <a><Button fullWidth>Checkout</Button></a>
                                </Link> */}
                <Stripe />
              </footer>
            </CartStyles>
          </>
        );
      }}
    </Composed>
  );
};

export default CartContainer;
export { LOCAL_STATE_QUERY, TOGGLE_CART_MUTATION };
