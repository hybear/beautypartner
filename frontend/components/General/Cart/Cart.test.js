import { MockedProvider } from 'react-apollo/test-utils';
import wait from 'waait';
import Cart, { LOCAL_STATE_QUERY } from './index'
import { CURRENT_USER_QUERY } from '../../User'
import { fakeUser, fakeCartItem } from '../../../lib/testUtils';

const mocks = [
    {
        request: { query: CURRENT_USER_QUERY},
        result: {
            data: { me: {
                ...fakeUser(),
                cart: [fakeCartItem()]
            }}
        }
    },
    {
        request: { query: LOCAL_STATE_QUERY },
        result: { data: {cartOpen: true } }
    }
]

describe('<Cart/>', () => {
    it('renders and matches snapshot', async () => {
        const wrapper = mount(
            <MockedProvider mocks={mocks}>
                <Cart/>
            </MockedProvider>
        )
        await wait(0);
        wrapper.update();
        expect(wrapper.find('header')).toMatchSnapshot()
        expect(wrapper.find('CartItem')).toHaveLength(1);
    });
});
