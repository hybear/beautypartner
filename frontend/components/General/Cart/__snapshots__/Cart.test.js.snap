// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cart/> renders and matches snapshot 1`] = `
<header>
  <styles__CloseButton
    onClick={[Function]}
  >
    <StyledComponent
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "styles__CloseButton-sc-6wf167-5",
            "isStatic": false,
            "lastClassName": "iyjIFn",
            "rules": Array [
              "color:",
              [Function],
              ";font-size:3rem;line-height:2rem;border:0;position:absolute;z-index:2;cursor:pointer;",
            ],
          },
          "displayName": "styles__CloseButton",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "styles__CloseButton-sc-6wf167-5",
          "target": "span",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
      onClick={[Function]}
    >
      <span
        className="styles__CloseButton-sc-6wf167-5 iyjIFn"
        onClick={[Function]}
      >
        Ã—
      </span>
    </StyledComponent>
  </styles__CloseButton>
  <Title>
    <StyledComponent
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "Title-sc-1er30d2-0",
            "isStatic": false,
            "lastClassName": "hOJAge",
            "rules": Array [
              "font-size:1.875em;margin:20px 0 40px;text-align:center;color:",
              [Function],
              ";",
            ],
          },
          "displayName": "Title",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "Title-sc-1er30d2-0",
          "target": "h2",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
    >
      <h2
        className="Title-sc-1er30d2-0 hOJAge"
      >
        Miss's
         Cart
      </h2>
    </StyledComponent>
  </Title>
  <p>
    You have 1 products in your cart
    .
  </p>
</header>
`;
